_posts/2019-01-26-generic-way-of-handling-loading-saving-and-validationerrors-in-angular.md

```diff
@@ -0,0 +1,410 @@
+tags: Angular
+cover: 'assets/images/cover/cover2.jpg'
+---
+
+# A generic way of handling loading-status, saving-status and validation errors in Angular
+
+When writing Angular applications there are always pieces of functionality that are being rewritten over and over again. 3 common usecases are:
+
+- Showing a *loading* status
+- Showing an *acting* status (whether the user is adding, updating or removing data)
+- Showing validation errors
+
+In this article we are going to implement a generic solution on how to fix these usecases.
+
+## A non generic way of loading, saving and handling validation errors in Angular.
+
+Before we jump to the solution, let's have a look at the impact of a non generic way of handling the previous called functionalities.
+A solution that is often used might look like the following: (keep in mind that this functionality has to be implemented over and over again for every component) 
+
+```typescript
+loading: boolean;
+acting: boolean;
+
+ngOnInit(): void {
+    this.loading = true;
+    const usersCompleted = false;
+    const citiesCompleted = false;
+    this.userService.fetch().subscribe(users => {
+        usersCompleted = true;
+        this.loading = usersCompleted && citiesCompleted;
+    });
+    this.citiesService.fetch().subscribe(cities => {
+        citiesCompleted = true;
+        this.loading = usersCompleted && citiesCompleted;
+    });
+}
+```
+We have to keep track of which call completes first because we can't set loading to false, if there is call still busy. What happens when there is an error? We would have to implement that as well.
+
+Now this code is only for fetching two lists of data, this becomes ugly pretty quickly and the worst thing about this is that we have to reimplement that for every component that does data fetching.
+
+It becomes even worse if we want to update, add and remove data. Imagine that we have to handle validation errors as well:
+
+```typescript
+remove(user): void {
+    // TODO: set acting to true
+    this.userService.remove(user).subscribe(res => {
+        // TODO: set acting to false
+    })
+}
+update(user): void {
+    // TODO: set acting to true
+    this.userService.update(user).subscribe(res => {
+        // TODO: set acting to false
+        // TODO: handle validation errors
+    })
+}
+add(user): void {
+    // TODO: set acting to true
+    this.userService.add(user).subscribe(res => {
+        // TODO: set acting to false
+        // TODO: handle validation errors
+    })
+}
+```
+
+To handle validation errors we have to check if the httpstatusCode is 400, manually map the data etc.
+
+These codesamples are in this article to prove a point. **It's dirty redundant logic that we have to implement over and over again**
+
+## Let's clean this up
+
+To achieve this we will use an angular *service* in combination with an angular *interceptor* and typescript *decorators*.
+
+The first thing we need is a `HttpStatusService` that exposes 3 observables:
```

Comment:
IMHO, service could be called HttpRequestProcess or Progress as it is shown that request is in progress